lib/Infer/AbstractInterpreter.cpp

```diff
@@ -715,10 +723,38 @@ namespace souper {
       Result.Zero.setHighBits(KB0.getBitWidth() - val.getActiveBits());
       break;
     }
-//   case FShl:
-//     return "fshl";
-//   case FShr:
-//     return "fshr";
+    case Inst::FShl: {
+      auto NewKB0 = concatKnownBits(KB0, KB1);
+      if (KB2.isConstant()) {
+        auto Shift = KB2.getConstant().urem(I->Width);
+        NewKB0.Zero = NewKB0.Zero.shl(Shift);
+        NewKB0.One = NewKB0.One.shl(Shift);
+        Result = NewKB0.trunc(I->Width);
```

Comment:
Are you sure this is correct? 

LangRef for fshl: 
"and the most significant bits are extracted to produce a result"

And `trunc` removes the most significant bits. LangRef for `trunc` says:
"The ‘trunc’ instruction truncates the high order bits in value "

Reply:
I have filed #631 for this so that we don't forget about it.